### PR TITLE
rework types, migrate parsing and add encoding

### LINKED ISSUE #43: Maybe `utf-8` is the best choise for solve decode error.
From: #5 #39 

My environment is
'Python': '3.9.13', 
'Platform': 'Windows-10-10.0.19044-SP0'
Call `locale.getpreferredencoding()` will return `cp936`,but decode error still occured.

PS: Another option is to support pytest.toml or pytest.yaml ?

### LINKED ISSUE #44: Is class IniConfig support char code option?
I use tox the testing tool for python in Windows 10 enviroment.
The problem is that class IniConfig tries to read tox.ini with cp932 encoding.
I read the code https://github.com/pytest-dev/iniconfig/blob/master/src/iniconfig/__init__.py#L51 . This not seems to have argument of encoding. Is this possible to set encoding?

* run command:

```ps
tox -c $Env:PROJECT_DIR/tox.ini -e py39
```

* bottom of stack trace

```
  File "c:\work\reni\git\ogs\.venv\lib\site-packages\py\_vendored_packages\iniconfig\__init__.py", line 54, in __init__
    tokens = self._parse(iter(f))
  File "c:\work\reni\git\ogs\.venv\lib\site-packages\py\_vendored_packages\iniconfig\__init__.py", line 82, in _parse
    for lineno, line in enumerate(line_iter):
UnicodeDecodeError: 'cp932' codec can't decode byte 0x81 in position 115: illegal multibyte sequence
```

### LINKED ISSUE #48: Specify encoding explictly in the `IniConfig` constructor
Hi!

Since Python 3.10 there's a new `EncodeWarning` that gets flagged when a file is opened without specifying an encoding. PR [#10328](https://github.com/pytest-dev/pytest/pull/10328) in pytest proposes adding this capability. For this to work, a bunch of calls need to be updated to explicitly specify `encoding="utf-8"` (or something similar – though UTF-8 seems like a sensible default).

Some of the tests, like [`test_ini_markers_whitespace`](https://github.com/pytest-dev/pytest/blob/9fbd67dd4b1baa6889dbb2073c17b85da39f80d9/testing/test_mark.py#L104), initialise an `IniConfig` object, whose constructor opens a file without specifying an encoding (see [here](https://github.com/pytest-dev/iniconfig/blob/1022b07e34bdb1fd378eaedee0bef8c98d77aaf4/src/iniconfig/__init__.py#L51)). This leads to the test failing.

Can we modify the constructor to use an explicit encoding when opening the file? If `iniconfig` doesn't support UTF-8 we can just use ASCII – so long as something is specified.

Thanks!

### PR DESCRIPTION
**Pull request by RonnyPfannschmidt**
* puts typing inline
* uses overloads to narrow types
* moves parsing into own file
* adds a encoding parameter that defaults to utf-8 - needs pytest crosscheck

fixes #44 
fixes #48 
addresses #5
fixes #43

### CODE REVIEW

**Review by nicoddemus — APPROVED** [review]
LGTM!

Also needs a CHANGELOG update to mention all the fixes, or at least #48?

File: MANIFEST.in [hunk 1]
```diff
@@ -2,4 +2,3 @@ include LICENSE
 include example.ini
```
[the remark this answers is not in the document]
Comment (RonnyPfannschmidt) [reply MANIFEST.in:2]: probably not anymore

**Comment by RonnyPfannschmidt** [comment]
i'll squash the fixup commits then prepare for more

File: CHANGELOG [hunk 2]
```diff
@@ -1,9 +1,11 @@
-Unreleased
+1.2.0 (Unreleased)
 ==========
 
-* add support for Python 3.6-3.10
+* add support for Python 3.6-3.11
```
Comment (nicoddemus) [inline CHANGELOG:4]: 3.6 is no longer supported I think.

File: src/iniconfig/__init__.py [hunk 3]
```diff
@@ -1,164 +1,175 @@
...
             yield name, self[name]
 
 
 class IniConfig:
-    def __init__(self, path, data=None):
+    path: Final[str]
+    sections: Final[Mapping[str, Mapping[str, str]]]
+
+    def __init__(
+        self, path: str, data: str | None = None, encoding: str = "utf-8"
```
Comment (inwaves) [inline src/iniconfig/__init__.py:74]: Just out of curiosity, is there any advantage to using a `Type | None` pattern over `Optional[Type]`? (I find the latter a little more readable – but I guess it's down to personal preference.)

Comment (RonnyPfannschmidt) [reply src/iniconfig/__init__.py:74]: Both mean the same, the one used there is the modern syntax and I prefer it

File: src/iniconfig/__init__.py [hunk 4]
```diff
@@ -1,164 +1,175 @@
...
+_D = TypeVar("_D")
+_T = TypeVar("_T")
 
 
 class SectionWrapper:
-    def __init__(self, config, name):
+    config: Final[IniConfig]
+    name: Final[str]
+
+    def __init__(self, config: IniConfig, name: str):
```
Comment (bluetech) [inline src/iniconfig/__init__.py:38]:
```suggestion
    def __init__(self, config: IniConfig, name: str) -> None:
```

File: src/iniconfig/__init__.py [hunk 5]
```diff
@@ -1,164 +1,175 @@
...
+    def lineof(self, name: str) -> int | None:
         return self.config.lineof(self.name, name)
 
-    def get(self, key, default=None, convert=str):
+    def get(
+        self,
+        key: str,
+        default: _D | None = None,
+        convert: Callable[[str], _T] | None = None,
+    ) -> _D | _T | str | None:
```
Comment (bluetech) [inline src/iniconfig/__init__.py:50]: If you're feeling charitable, can add a bunch of overloads here to give better return type when `default` and/or `convert` are None.

File: src/iniconfig/__init__.py [hunk 6]
```diff
@@ -1,164 +1,175 @@
...
 
-    def __iter__(self):
-        section = self.config.sections.get(self.name, [])
+    def __iter__(self) -> Iterator[str]:
+        section: Mapping[str, str] = self.config.sections.get(self.name, {})
 
-        def lineof(key):
-            return self.config.lineof(self.name, key)
+        def lineof(key: str) -> int:
+            return self.config.lineof(self.name, key)  # type: ignore
```
Comment (bluetech) [inline src/iniconfig/__init__.py:60]: I recommend always adding the error code to `type: ignores`, e.g. `type: ignore[arg-type]` (just an example, I'm not sure what is the error code in this case)

File: src/iniconfig/__init__.py (hunk 3 above)
Comment (bluetech) [inline src/iniconfig/__init__.py:74]: Maybe make path `str | os.PathLike[str]` (and use `os.path` for the "convenience" part below)

File: src/iniconfig/exceptions.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+if TYPE_CHECKING:
+    from typing_extensions import Final
+
+
+class ParseError(Exception):
+    path: Final[str]
+    lineno: Final[int]
+    msg: Final[str]
+
+    def __init__(self, path: str, lineno: int, msg: str):
```
Comment (bluetech) [inline src/iniconfig/exceptions.py:13]:
```suggestion
    def __init__(self, path: str, lineno: int, msg: str) -> None:
```

File: src/iniconfig/exceptions.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+        Exception.__init__(self, path, lineno, msg)
```
Comment (bluetech) [inline src/iniconfig/exceptions.py:14]:
```suggestion
        super().__init__(path, lineno, msg)
```

**Pull request merged**